File: mz_wuwa_cal/mz_wuwa_cal.go
package mzwuwacal

type Attacker struct {
	// Basic info of the attacker
	name        string
	level       int
	element     string
	attack_Type int

	// true damage of the attacker
	true_dmg float64

	// Base damage of the attacker
	base_dmg    int64
	ability_dmg float64
	// total attack of the attacker
	total_atk     int64
	base_atk      int64
	atk_buff      float64
	flat_atk_buff int64
	flat_atk      int64

	// DMG Bonus of the attacker
	bonus_multiplier float64
	dmg_bonus        float64
	dmg_amplify      float64
	special_dmg      float64
	crit_dmg         float64
}

type Target struct {
	// Basic info of the target
	name    string
	def     int64
	level   int
	element string

	// all resistances of the target
	all_resistances    float64
	base_element_res   float64
	def_multiplier     float64
	dmg_reduce         float64
	element_dmg_reduce float64

	// negative resistances of the target
	def_ignore         float64 // <-> def
	dmg_taken          float64 // <-> dmg_reduce
	element_res_ignore float64 // <-> base_element_res
}

// Damage Calculation
func GetTrueDamage(a *Attacker, t *Target) float64 {
	return GetBaseDamage(a) * GetAllResistances(t) * a.bonus_multiplier
}

func GetBaseDamage(a *Attacker) float64 {
	return a.ability_dmg * GetTotalAttack(a)
}

func GetTotalAttack(a *Attacker) float64 {
	return float64(a.base_atk)*(1+a.atk_buff) + float64(a.flat_atk) + float64(a.flat_atk_buff)
}

// Resistances Calculation
func GetAllResistances(t *Target) float64 {
	return GetBaseElementRes(t) * t.def_multiplier * t.dmg_reduce * t.element_dmg_reduce
}

func GetBaseElementRes(t *Target) float64 {

	res_total := t.base_element_res - t.element_res_ignore

	if res_total < 0 {
		return 1 - (res_total / 2)
	} else if res_total < 0.8 {
		return 1 - res_total
	} else { // res_total => 0.8
		return 1 / (1 + (5 * res_total))
	}

}

File: wuwa_cal_rust/src/mz_wuwa_cal.rs
#[derive(Debug)]
pub struct Attacker {
    pub name: String,
    pub level: i32,
    pub ability_dmg: f64,
    pub total_atk: i64,
    pub crit: bool,
    pub element_dmg_bonus: f64,
    pub attack_type_dmg_bonus: f64,
    pub dmg_amplify: f64,
    pub crit_dmg: f64,
    pub def_ignore: f64,
    pub element_res_ignore: f64,
    pub special_dmg: f64,
}

#[derive(Debug)]
pub struct Target {
    pub name: String,
    pub level: i32,
    pub base_element_res: f64,
    pub dmg_taken: f64,
    pub dmg_res_base: f64,
    pub dmg_res_additional: f64,
    pub element_res_base: f64,
    pub element_res_additional: f64,
    pub dmg_amplify: f64,
}

pub fn get_true_damage(a: &Attacker, t: &mut Target) -> f64 {
    get_base_damage(a) * get_all_resistances(a, t) * get_bonus_multipliers(a, t)
}

pub fn get_base_damage(a: &Attacker) -> f64 {
    a.ability_dmg * a.total_atk as f64
}

pub fn get_all_resistances(a: &Attacker, t: &mut Target) -> f64 {
    get_base_element_res(a, t)
        * get_def_multiplier(a, t)
        * get_dmg_reduce(t)
        * get_element_reduce(t)
}

pub fn get_base_element_res(a: &Attacker, t: &Target) -> f64 {
    let res_total = t.base_element_res - a.element_res_ignore;

    if res_total < 0.0 {
        1.0 - (res_total / 2.0)
    } else if res_total < 0.8 {
        1.0 - res_total
    } else {
        1.0 / (1.0 + (5.0 * res_total))
    }
}

pub fn get_def_multiplier(a: &Attacker, t: &Target) -> f64 {
    let target_def = (8 * t.level + 792) as f64;
    let attacker_multiplier = (800 + 8 * a.level) as f64;
    attacker_multiplier / (attacker_multiplier + target_def * (1.0 - a.def_ignore))
}

pub fn get_dmg_reduce(t: &mut Target) -> f64 {
    t.dmg_res_base = 0.0;
    t.dmg_res_additional = 0.0;
    1.0 - (t.dmg_res_base + t.dmg_res_additional) + t.dmg_taken
}

pub fn get_element_reduce(t: &mut Target) -> f64 {
    t.dmg_res_base = 0.0;
    t.dmg_res_additional = 0.0;
    1.0 - (t.element_res_base + t.element_res_additional)
}

pub fn get_bonus_multipliers(a: &Attacker, t: &mut Target) -> f64 {
    if a.crit {
        get_dmg_bonus(a) * get_dmg_amplify(a, t) * get_special_dmg(a) * a.crit_dmg
    } else {
        get_dmg_bonus(a) * get_dmg_amplify(a, t) * get_special_dmg(a)
    }
}

pub fn get_dmg_bonus(a: &Attacker) -> f64 {
    1.0 + a.element_dmg_bonus + a.attack_type_dmg_bonus
}

pub fn get_dmg_amplify(a: &Attacker, t: &mut Target) -> f64 {
    t.dmg_amplify = 0.0;
    1.0 + a.dmg_amplify + t.dmg_amplify
}

pub fn get_special_dmg(a: &Attacker) -> f64 {
    1.0 + a.special_dmg
}
